File: examples/todo_app/nim_todo/tests/test_todo_store.nim
## Port of c_todo/tests/test_todo_store.c
import db_connector/sqlite3 as s3
import todo_store

var gFail = false

proc fail(msg: string) =
  stderr.writeLine "test_todo_store: ", msg
  gFail = true

proc main =
  gFail = false
  let s = storeOpen("")
  if s.isNil:
    fail("todo_store_open(\"\")")
    quit QuitFailure

  var cat1: int64 = 0
  if storeCategoryInsert(s, "Work", "#ff00aa", cat1) != s3.SQLITE_OK:
    fail("category insert")
  if cat1 <= 0:
    fail("category id")

  var dup: int64 = 0
  if storeCategoryInsert(s, "Work", "#000000", dup) == s3.SQLITE_OK:
    fail("duplicate category name")

  var catCount = 0
  let rc1 = storeCategoriesForeach(s) do (r: TodoStoreCategoryRow):
    inc catCount
  if rc1 != s3.SQLITE_OK:
    fail("categories foreach")
  if catCount != 1:
    fail("category count")

  var tid: int64 = 0
  if storeTodoInsert(s, "Task one", "", "pending", "high", true, cat1, false, 0, false, 0, tid) !=
      s3.SQLITE_OK:
    fail("todo insert")
  if tid <= 0:
    fail("todo id")

  var todoCount = 0
  let rc2 = storeTodosForeach(s) do (r: TodoStoreTodoRow):
    inc todoCount
  if rc2 != s3.SQLITE_OK:
    fail("todos foreach")
  if todoCount != 1:
    fail("todo foreach count")

  if storeTodoUpdateFull(s, tid, "Task one", "desc", "in_progress", "low", false, 0, false, 0, false, 0, 0) !=
      s3.SQLITE_OK:
    fail("todo update")

  if storeTodoUpdateFull(s, tid, "Task one", "desc", "archived", "low", false, 0, false, 0, false, 0, 0) !=
      s3.SQLITE_OK:
    fail("todo update archived")

  if storeTodoDeleteById(s, tid) != s3.SQLITE_OK:
    fail("todo delete")
  todoCount = 0
  let rc3 = storeTodosForeach(s) do (r: TodoStoreTodoRow):
    inc todoCount
  if rc3 != s3.SQLITE_OK:
    fail("todos foreach 2")
  if todoCount != 0:
    fail("todo count after delete")

  if storeCategoryDeleteById(s, cat1) != s3.SQLITE_OK:
    fail("category delete")
  catCount = 0
  let rc4 = storeCategoriesForeach(s) do (r: TodoStoreCategoryRow):
    inc catCount
  if rc4 != s3.SQLITE_OK:
    fail("cat foreach 2")
  if catCount != 0:
    fail("categories empty")

  var bad: int64 = 0
  if storeCategoryInsert(s, "", "#fff", bad) == s3.SQLITE_OK:
    fail("empty category name")
  if storeTodoInsert(s, "", "x", "pending", "medium", false, 0, false, 0, false, 0, tid) == s3.SQLITE_OK:
    fail("empty title")

  storeClose(s)
  quit(if gFail: QuitFailure else: QuitSuccess)

when isMainModule:
  main()

File: examples/todo_app/rust_todo/tests/test_todo_store.rs
// Port of c_todo/tests/test_todo_store.c / nim_todo/tests/test_todo_store.nim

use rusqlite::ffi::SQLITE_OK;
use rust_todo::todo_store::{
    store_categories_foreach, store_category_delete_by_id, store_category_insert, store_close,
    store_open, store_todo_delete_by_id, store_todo_insert, store_todo_update_full,
    store_todos_foreach, TodoStoreCategoryRow, TodoStoreTodoRow,
};

#[test]
fn todo_store_lifecycle() {
    let s = store_open("").expect("todo_store_open(\"\")");

    let mut cat1 = 0i64;
    assert_eq!(store_category_insert(&s, "Work", "#ff00aa", &mut cat1), SQLITE_OK);
    assert!(cat1 > 0, "category id");

    let mut dup = 0i64;
    assert_ne!(
        store_category_insert(&s, "Work", "#000000", &mut dup),
        SQLITE_OK,
        "duplicate category name"
    );

    let mut cat_count = 0;
    assert_eq!(
        store_categories_foreach(&s, |_r: TodoStoreCategoryRow| {
            cat_count += 1;
        }),
        SQLITE_OK
    );
    assert_eq!(cat_count, 1, "category count");

    let mut tid = 0i64;
    assert_eq!(
        store_todo_insert(
            &s,
            "Task one",
            "",
            "pending",
            "high",
            true,
            cat1,
            false,
            0,
            false,
            0,
            &mut tid,
        ),
        SQLITE_OK
    );
    assert!(tid > 0, "todo id");

    let mut todo_count = 0;
    assert_eq!(
        store_todos_foreach(&s, |_r: TodoStoreTodoRow| {
            todo_count += 1;
        }),
        SQLITE_OK
    );
    assert_eq!(todo_count, 1, "todo foreach count");

    assert_eq!(
        store_todo_update_full(
            &s,
            tid,
            "Task one",
            "desc",
            "in_progress",
            "low",
            false,
            0,
            false,
            0,
            false,
            0,
            0,
        ),
        SQLITE_OK
    );

    assert_eq!(
        store_todo_update_full(
            &s,
            tid,
            "Task one",
            "desc",
            "archived",
            "low",
            false,
            0,
            false,
            0,
            false,
            0,
            0,
        ),
        SQLITE_OK
    );

    assert_eq!(store_todo_delete_by_id(&s, tid), SQLITE_OK);
    todo_count = 0;
    assert_eq!(
        store_todos_foreach(&s, |_r: TodoStoreTodoRow| {
            todo_count += 1;
        }),
        SQLITE_OK
    );
    assert_eq!(todo_count, 0, "todo count after delete");

    assert_eq!(store_category_delete_by_id(&s, cat1), SQLITE_OK);
    cat_count = 0;
    assert_eq!(
        store_categories_foreach(&s, |_r: TodoStoreCategoryRow| {
            cat_count += 1;
        }),
        SQLITE_OK
    );
    assert_eq!(cat_count, 0, "categories empty");

    let mut bad = 0i64;
    assert_ne!(
        store_category_insert(&s, "", "#fff", &mut bad),
        SQLITE_OK,
        "empty category name"
    );
    assert_ne!(
        store_todo_insert(
            &s,
            "",
            "x",
            "pending",
            "medium",
            false,
            0,
            false,
            0,
            false,
            0,
            &mut tid,
        ),
        SQLITE_OK,
        "empty title"
    );

    store_close(s);
}
